apps/publikator/components/Files/Upload.js

```diff
@@ -0,0 +1,137 @@
+import { useState } from 'react'
+import { gql, useMutation } from '@apollo/client'
+import Dropzone from 'react-dropzone'
+import axios from 'axios'
+import { css } from 'glamor'
+
+import { ProgressCircle } from '@project-r/styleguide'
+
+import { RepoFile } from '../../lib/graphql/fragments'
+
+const UPLOAD_BEGIN = gql`
+  mutation repoFileUploadBegin($repoId: ID!, $name: String!) {
+    repoFileUploadBegin(repoId: $repoId, name: $name) {
+      ...RepoFile
+    }
+  }
+
+  ${RepoFile}
+`
+
+const UPLOAD_COMMIT = gql`
+  mutation repoFileUploadCommit($id: ID!) {
+    repoFileUploadCommit(id: $id) {
+      ...RepoFile
+    }
+  }
+
+  ${RepoFile}
+`
+
+const styles = {
+  dropzone: css({
+    height: '15rem',
+    display: 'flex',
+    justifyContent: 'center',
+    alignItems: 'center',
+    margin: '2rem 0',
+  }),
+}
+
+export default ({ repoId }) => {
+  const [uploading, setUploading] = useState(false)
+  const [progress, setProgress] = useState(100)
+
+  const [uploadBegin] = useMutation(UPLOAD_BEGIN)
+  const [uploadCommit] = useMutation(UPLOAD_COMMIT)
+
+  // @TODO handle rejected files
+  const onDrop = async (accepted) => {
+    setUploading(true)
+    setProgress(0)
+
+    try {
+      const dropTotal = accepted.reduce((size, file) => size + file.size, 0)
+      let dropLoaded = 0
+
+      for await (const file of accepted) {
+        const { data } = await uploadBegin({
+          variables: { repoId, name: file.name },
+          refetchQueries: ['getFiles'],
+          awaitRefetchQueries: true,
+        })
+
+        const id = data?.repoFileUploadBegin?.id
+        const url = data?.repoFileUploadBegin?.url
+
+        const res = await axios(url, {
+          method: 'PUT',
+          headers: {
+            'Content-Type': file.type,
+          },
+          data: file,
+          onUploadProgress: (progressEvent) => {
+            const { loaded: fileLoaded, total: fileTotal } = progressEvent
+
+            if (fileTotal) {
+              setProgress(
+                Math.round(((dropLoaded + fileLoaded) / dropTotal) * 100),
+              )
+
+              if (fileLoaded >= fileTotal) {
+                dropLoaded += fileLoaded
+              }
+            }
+          },
+        })
+
+        if (!(res.status >= 200 && res.status < 300)) {
```

Comment:
so th api error comes up here and the console error below, right? why do we exclude other error codes (40x, 500)?

Reply:
It throws an error whenever code is outside 200 to 299. 4xx and 5xx are included.